src/lem/interpreter.rs

```diff
@@ -1,206 +1,253 @@
+    fn run<F: LurkField>(
         &self,
-        input: [Ptr<F>; 3],
         store: &mut Store<F>,
-    ) -> Result<Witness<F>> {
-        // key/val pairs on this map should never be overwritten
-        let mut ptrs = HashMap::default();
-        ptrs.insert(self.input[0].clone(), input[0]);
-        insert_into_ptrs(&mut ptrs, self.input[1].clone(), input[1])?;
-        insert_into_ptrs(&mut ptrs, self.input[2].clone(), input[2])?;
-        let mut output = None;
-        let mut stack = vec![&self.lem_op];
-        while let Some(op) = stack.pop() {
-            match op {
-                LEMOP::Null(tgt, tag) => {
-                    let tgt_ptr = Ptr::null(*tag);
-                    insert_into_ptrs(&mut ptrs, tgt.name().clone(), tgt_ptr)?;
-                }
-                LEMOP::Hash2(tgt, tag, src) => {
-                    let src_ptr1 = src[0].get_ptr(&ptrs)?;
-                    let src_ptr2 = src[1].get_ptr(&ptrs)?;
-                    let tgt_ptr = store.intern_2_ptrs(*tag, src_ptr1, src_ptr2);
-                    insert_into_ptrs(&mut ptrs, tgt.name().clone(), tgt_ptr)?;
-                }
-                LEMOP::Hash3(tgt, tag, src) => {
-                    let src_ptr1 = src[0].get_ptr(&ptrs)?;
-                    let src_ptr2 = src[1].get_ptr(&ptrs)?;
-                    let src_ptr3 = src[2].get_ptr(&ptrs)?;
-                    let tgt_ptr = store.intern_3_ptrs(*tag, src_ptr1, src_ptr2, src_ptr3);
-                    insert_into_ptrs(&mut ptrs, tgt.name().clone(), tgt_ptr)?;
-                }
-                LEMOP::Hash4(tgt, tag, src) => {
-                    let src_ptr1 = src[0].get_ptr(&ptrs)?;
-                    let src_ptr2 = src[1].get_ptr(&ptrs)?;
-                    let src_ptr3 = src[2].get_ptr(&ptrs)?;
-                    let src_ptr4 = src[3].get_ptr(&ptrs)?;
-                    let tgt_ptr = store.intern_4_ptrs(*tag, src_ptr1, src_ptr2, src_ptr3, src_ptr4);
-                    insert_into_ptrs(&mut ptrs, tgt.name().clone(), tgt_ptr)?;
-                }
-                LEMOP::Unhash2(tgts, src) => {
-                    let src_ptr = src.get_ptr(&ptrs)?;
-                    let Some(idx) = src_ptr.get_index2() else {
-                        bail!(
-                            "{} isn't a Tree2 pointer",
-                            src.name()
-                        );
-                    };
-                    let Some((a, b)) = store.fetch_2_ptrs(idx) else {
-                        bail!("Couldn't fetch {}'s children", src.name())
-                    };
-                    insert_into_ptrs(&mut ptrs, tgts[0].name().clone(), *a)?;
-                    insert_into_ptrs(&mut ptrs, tgts[1].name().clone(), *b)?;
+        bindings: &mut HashMap<MetaPtr, Ptr<F>>,
+        preimages: &mut Preimages<F>,
+    ) -> Result<()> {
+        match self {
+            LEMOP::Null(tgt, tag) => {
+                let tgt_ptr = Ptr::null(*tag);
+                bind(bindings, tgt.clone(), tgt_ptr)
```

Comment:
If I understand this approach, the effective equivalent of the old `Witness` struct is the `bindings` map — and you are storing a variable-value binding for *every single intermediate value* in every frame of evaluation. This is the opposite of the approach we should be taking, which is to minimize the non-deterministic hint/witnesses used to construct the circuit. In some cases, knowledge of preimages is unavoidable — but requiring everything is a step all the way in the opposite direction from what we are trying to do.